### DISCUSSION TITLE
Generic Type - Find max

### DISCUSSION THREAD

**Discussion by CocDap** [openedu101/Rust-bootcamp-2024#16] [Q&A]
Input: 
Cho 1 collection (có thể là vector, array ) 

Output:
Tìm số lớn nhất 

```rust
fn find_max<T>(collection: &[T]) -> Option<&T> {
    todo!()
}
```

Run test case : 

```rust
    #[test]
    fn test_find_max_with_numbers() {
        let numbers = vec![1, 5, 3, 8, 2];
        assert_eq!(find_max(&numbers), Some(&8));
    }

    #[test]
    fn test_find_max_with_empty_collection() {
        let empty: Vec<i32> = Vec::new();
        assert_eq!(find_max(&empty), None);
    }
```

**Comment by tihii8898** [comment]
Thêm `T: Ord` để impl trait Ord. Từ đó có thể dùng các hàm so sánh do Rust có sẳn trong `iter`.
Cái này mình hiểu theo cách của bản thân nên các anh và các bạn thấy chưa đúng cho mình xin ý kiến để cả thiện ạ. Cảm ơn mọi người.

```
pub fn find_max<T: Ord>(collection: &[T]) -> Option<&T> {
    if collection.is_empty() {
        return None;
    }
    let max = collection.iter().max().unwrap();

    Some(max)
}
```

**Reply by CocDap** [reply]
Cái này bạn impl hơi bị dư nhé 
`vì collection.iter().max()` đã trả về None hoặc Some(_) nhé ( https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max)

Còn bạn thêm `trait bound` `Ord` là đúng rùi đó

**Reply by CocDap** [reply]
<img width="1090" alt="image" src="https://github.com/openedu101/Rust-bootcamp-2024/assets/48429826/7517ac53-788b-4840-990a-cca74fccd209">

Vì method `max` này có constraint bởi `Ord`, `Ord` dùng để liên quan tới thứ tự ví dụ so sánh lớn hơn , bé hơn, bằng, max, min

**Reply by tihii8898** [reply]
Em cảm ơn anh nhiều
